ls1intum/Artemis
# [Bugfix] Fix issues for team assessments

### Checklist

### Motivation and Context


Recently, errors in the (team) assessment view were reported: Solves issue #3389
Solves sentry issue [ARTEMIS-10Y ](https://sentry.ase.in.tum.de/organizations/artemis/issues/1068). 
When solving these problems I had to make changes that are described in detail below.

### Description


Following changes were made:

1. For exercises which have submissions but are not due yet, instructors were allowed to view the submission and even assess them. The **warning on the submission page**, is now also shown on the team page and within the assessment editor. (This does not apply to programming exercises, they can't be assessed before due date)

e.g. https://artemistest.ase.in.tum.de/course-management/477/file-upload-exercises/7468/submissions
![image](https://user-images.githubusercontent.com/41366522/119448834-27485b80-bd32-11eb-962a-69d07943c81a.png)

e.g. https://artemistest.ase.in.tum.de/course-management/477/file-upload-exercises/7468/submissions/437267/assessment
![image](https://user-images.githubusercontent.com/41366522/119448805-1e578a00-bd32-11eb-83e2-50189657142d.png)

2. Recently, errors during the assessment of team exercises occurred. Instructors AND tutors could assess teams/submissions when navigating on the team page. Assessments could be started for exercises without a submitted solution and also for exercises before due date, independent of the role. When submissions were then saved/submitted errors occurred. The current behavior does also not **comply with the assessment rules implemented for individual assessment** -> there, tutors could not assess exercises before due date for example.

e.g. https://artemistest.ase.in.tum.de/course-management/477/programming-exercises/7643/teams/395
Buttons for starting/open/continuing the assessment were disabled according to the following logic:

For all team exercises:
- Exercises should have a submission with `submitted=true`

For team exercises with due date:
- Instructors can assess submission (except for programming exercise), but should be warned on the team page and in the assessment editor.
- Tutors can assess submissions only after due date

An example situation for different exercises in different states:
Tutor view vs. Instructor view (red: no assessment possible -> button disabled, green: assessment possible -> button enabled)


<img width="2555" alt="Bildschirmfoto 2021-05-25 um 08 47 06" src="https://user-images.githubusercontent.com/41366522/119452453-adff3780-bd36-11eb-9226-2f1626ab5c1e.png">


3. The Team Participation Table was reduced, as there were misleading information and the whole component was not tested


### Steps for Testing


_Background Info_:
- Team exercises have to be created as such, choose the mode TEAM in the exercise edit page
- You need to add teams after creating an exercise -> create a new team or import a team of other exercises, also choose a tutor for the team; do this once before testing and use the same team (via import from exercise) for all team exercise you create for testing
- Navigation to the team page: search a team exercise in course management, (1) click on the teams button, (2) click on the link behind the team name

(1)
<img width="1380" alt="Bildschirmfoto 2021-05-25 um 10 16 44" src="https://user-images.githubusercontent.com/41366522/119463838-6e3e4d00-bd42-11eb-82e5-c01d9033a888.png">

(2)
<img width="1426" alt="Bildschirmfoto 2021-05-25 um 10 18 10" src="https://user-images.githubusercontent.com/41366522/119463942-8b731b80-bd42-11eb-9d7a-469aaff471d6.png">

1. Log in to Artemis as instructor
2. Recreate the situations described in the description above
3. Check if the assessment action buttons are enabled/disabled accordingly

1. Log in to Artemis as tutor for the exercise
2. Recreate the situations described in the description above
3. Check if the assessment action buttons are enabled/disabled accordingly

-> send me a message, if something is unclear or leave comment

### Test Coverage

team-participation-table.component.ts 92%
